### PR TITLE
Observation normalization

### PR DESCRIPTION
**Pull request by le-horizon**
This change
1) makes ObservationNormalizer update stats during training, not unroll.
2) sets default parameters for the best performance of HER.
3) introduces alf.layers.Module to replace nn.Module so that plotting network outputs can be made easier (to distinguish among the three possible modes the networks can be invoked: rollout, replay and eval).
plus some small refactors.

Thanks,
Le

### CODE REVIEW

File: alf/utils/normalizers.py [hunk 1]
```diff
@@ -81,6 +82,38 @@ def update(self, tensor):
         self._mean_averager.update(tensor)
         sqr_tensor = alf.nest.map_structure(math_ops.square, tensor)
         self._m2_averager.update(sqr_tensor)
+        if alf.summary.should_summarize_output():
```
Comment (hnyu) [inline alf/utils/normalizers.py]: Why is this summary controlled by summarize_output?

File: alf/nest/nest.py [hunk 2]
```diff
@@ -258,6 +258,37 @@ def extract_any_leaf_from_nest(nest):
                 return ret
 
 
+def all_leaves_with_path_from_nest(nest, path=""):
```
Comment (hnyu) [inline alf/nest/nest.py]: This will be inefficient. Should be in C++. A better way is to write a function called map_structure_with_path, and the lambda function is expected to take an extra path argument. It will throw an error if the structure doesn't have fields.

**Comment by emailweixu** [comment]
should condition the calling of summary functions on should_record_summaries() so that there is no waste on preparing the data for summary (e.g. norm(), mean(), etc). ALF already has a lot of overhead. We should be careful to avoid adding more.

**Comment by le-horizon** [comment]
> should condition the calling of summary functions on should_record_summaries() so that there is no waste on preparing the data for summary (e.g. norm(), mean(), etc). ALF already has a lot of overhead. We should be careful to avoid adding more.

Good point.  Forgot to do that for a few places in replay_buffer.  Done.

File: alf/utils/common.py [hunk 3]
```diff
@@ -441,7 +442,7 @@ def forward(self, observation):
         """Normalize a given observation. If during unroll, then first update
         the normalizer. The normalizer won't be updated in other circumstances.
         """
-        if not is_training():
+        if is_training():
```
Comment (hnyu) [inline alf/utils/common.py]: Probably it's better to add an option for the user to select when to update statistics

File: alf/utils/normalizers.py [hunk 4]
```diff
@@ -81,6 +82,38 @@ def update(self, tensor):
...
+            suffix = self.exe_mode_str()
+
+            def _summary(name, val):
+                if val.shape[0] < 30:
```
Comment (hnyu) [inline alf/utils/normalizers.py]: Use a variable to hold this constant 30

File: alf/layers.py [hunk 5]
```diff
@@ -100,8 +109,41 @@ def unflatten(self, tensor):
                      tuple(tensor.shape[1:])))
 
 
+class Module(nn.Module):
+    """Alf's nn.Module wrapper"""
```
Comment (hnyu) [inline alf/layers.py]: What's the purpose of having this new Module class?

File: alf/layers.py [hunk 6]
```diff
@@ -20,13 +20,22 @@
...
+import alf
 from alf.initializers import variance_scaling_init
 from alf.nest.utils import get_outer_rank
 from alf.tensor_specs import TensorSpec
 from alf.utils import common
 from alf.utils.math_ops import identity
 
 
+def reduce_along_batch_dims(x, mean, op):
+    spec = TensorSpec.from_tensor(mean)
```
Comment (hnyu) [inline alf/layers.py]: from_tensor by default will only treat dim=0 as the batch dim. Then there is no need to use BatchSquash to flatten the tensor x. Generally I feel that this function is not that necessary.

File: alf/algorithms/agent.py [hunk 7]
```diff
@@ -127,6 +127,7 @@ def __init__(self,
         rl_algorithm = rl_algorithm_cls(
             observation_spec=rl_observation_spec,
             action_spec=action_spec,
+            config=config,
```
Comment (hnyu) [inline alf/algorithms/agent.py]: Why do we need config here? usually passing a config means that this algorithm needs to create a replay buffer or set rollout state stuff.

**Review by hnyu — CHANGES_REQUESTED** [review]

File: alf/algorithms/agent.py (hunk 7 above)
Comment (le-horizon) [reply alf/algorithms/agent.py]: Good catch, this is only meant for updating in unroll, where a separate replay buffer for unroll experience is created in DdpgAlgorithm.

File: alf/layers.py (hunk 5 above)
Comment (le-horizon) [reply alf/layers.py]: This is to replace the nn.Module in alf for all custom code we need to have.  Here, it tracks the mode of rollout vs replay (during training).  It is also responsible for generating the mode string for tensorboard plotting purposes.

File: alf/utils/normalizers.py (hunk 1 above)
Comment (le-horizon) [reply alf/utils/normalizers.py]: Good point.  Using should_record_summary.

File: alf/utils/normalizers.py (hunk 4 above)
Comment (le-horizon) [reply alf/utils/normalizers.py]: Added a const: MAX_DIMS_TO_OUTPUT

File: alf/nest/nest.py (hunk 2 above)
Comment (le-horizon) [reply alf/nest/nest.py]: Added comment about efficiency.  This is used for tb summary, so let's leave it as is?  Also for display purposes lists are allowed currently, and assumed to have its index as path, e.g. goal.0, achieved_goal.1 etc..

File: alf/utils/common.py (hunk 3 above)
Comment (le-horizon) [reply alf/utils/common.py]: For unroll, we need to do it with HER sampling, which means we probably cannot reuse this class.

For the non-her case, we also need to fix the case where test exp should be excluded.

Added a TODO.

File: alf/layers.py (hunk 6 above)
Comment (le-horizon) [reply alf/layers.py]: mean is actually the tensor that has dim=0 as the batch dim, while x can include batch_size, batch_length two batch dims.

This function is modeled after the normalize_along_batch_dims function below it.

**Review by le-horizon — COMMENTED** [review]
@hnyu thanks for the quick and careful review.
I fixed or answered to all your comments.  PTAL.

Thanks,
Le

File: alf/algorithms/rl_algorithm.py [hunk 8]
```diff
@@ -452,8 +452,10 @@ def unroll(self, unroll_length):
             # to store it in replay buffers
             transformed_time_step = transformed_time_step._replace(
                 untransformed=time_step)
+            self.rollout()
```
Comment (emailweixu) [inline alf/algorithms/rl_algorithm.py]: We should avoid adding overhead like this. Recursive call can be expensive. Perhaps modify common.mark_training to handle all three modes.

File: alf/networks/actor_networks.py [hunk 9]
```diff
@@ -140,18 +140,20 @@ def forward(self, observation, state=()):
...
+            layer.summary_name = ""
             action = spec_utils.scale_to_spec(action, spec)
 
             if alf.summary.should_summarize_output():
-                a_name = ('summarize_output/' + self.name + '.action_layer.' +
-                          str(i) + '.action.output_norm')
-                if not self.training:
-                    a_name += ".eval"
-                a = action
+                a_name = (
```
Comment (emailweixu) [inline alf/networks/actor_networks.py:155]: should_summarize_output() will always return True if config.summarize_output is True. But summary is only performed when should_record_summaries() return True. Need to avoid the overhead by conditioning on should_record_summaries() too.

File: alf/layers.py [hunk 10]
```diff
@@ -100,8 +109,41 @@ def unflatten(self, tensor):
...
 
+class Module(nn.Module):
+    """Alf's nn.Module wrapper"""
+
+    def __init__(self):
+        super().__init__()
+        self.rollingout = False
+        self.summary_name = ""
+
+    def rollout(self, mode=True):
```
Comment (emailweixu) [inline alf/layers.py]: comment

File: alf/layers.py [hunk 11]
```diff
@@ -100,8 +109,41 @@ def unflatten(self, tensor):
                      tuple(tensor.shape[1:])))
 
 
+class Module(nn.Module):
+    """Alf's nn.Module wrapper"""
+
+    def __init__(self):
+        super().__init__()
+        self.rollingout = False
```
Comment (emailweixu) [inline alf/layers.py]: self._rollingout

File: alf/layers.py [hunk 12]
```diff
@@ -100,8 +109,41 @@ def unflatten(self, tensor):
...
+        self.summary_name = ""
```
Comment (emailweixu) [inline alf/layers.py]: self._summary_name
and use commented property accessor/setter for it since it's publicly accessible

**Review by emailweixu — CHANGES_REQUESTED** [review]
In general, try not to make existing function slower if possible.

File: alf/nest/nest.py (hunk 2 above)
Comment (hnyu) [reply alf/nest/nest.py]: OK we can support lists. I was thinking of some function like this:

```
map_structure_with_path(lambda t1, t2, path: summary_fn(path, t1+t2), tensor1, tensor2)
```

It basically takes a lambda function with an additional 'path' argument. This path argument will be automatically generated according to the current leaf visited by this map structure function.

The whole purpose of all_leaves_with_path_from_nest is to be used by another loop through all leaves. We can just combine them into one.

File: alf/utils/common.py (hunk 3 above)
Comment (hnyu) [reply alf/utils/common.py]: Might just set the flag to be one of "train", "unroll", and "eval"?

File: alf/layers.py (hunk 6 above)
Comment (hnyu) [reply alf/layers.py]: I think it's for specific purpose so probably it should be an inner function in whatever calls it and prefixed by '_'.

File: alf/algorithms/agent.py (hunk 7 above)
Comment (hnyu) [reply alf/algorithms/agent.py]: I'm still confused. If only config is passed but not env, then no replay buffer will be created?

File: alf/layers.py (hunk 5 above)
Comment (hnyu) [reply alf/layers.py]: Generally let's be conservative about changing some very basic module or building block because it will usually bring in efficiency issues or code structure issues. If there is some other way like common.mark_training to achieve the same purpose them we should try to avoid this.

File: alf/algorithms/rl_algorithm.py (hunk 8 above)
Comment (le-horizon) [reply alf/algorithms/rl_algorithm.py]: Good point.  The logic is cleaner this way.

File: alf/layers.py (hunk 6 above)
Comment (le-horizon) [reply alf/layers.py]: Good point.

File: alf/layers.py (hunk 5 above)
Comment (le-horizon) [reply alf/layers.py]: Good points.  Done.

File: alf/layers.py (hunk 11 above)
Comment (le-horizon) [reply alf/layers.py]: no longer needed.

File: alf/layers.py (hunk 12 above)
Comment (le-horizon) [reply alf/layers.py]: no longer needed.

File: alf/layers.py (hunk 10 above)
Comment (le-horizon) [reply alf/layers.py]: no longer needed.

File: alf/networks/actor_networks.py (hunk 9 above)
Comment (le-horizon) [reply alf/networks/actor_networks.py:155]: good point.

File: alf/utils/common.py (hunk 3 above)
Comment (le-horizon) [reply alf/utils/common.py]: Done.

File: alf/nest/nest.py (hunk 2 above)
Comment (le-horizon) [reply alf/nest/nest.py]: Good point.  Done.

**Review by le-horizon — COMMENTED** [review]
Thanks for the very good points guys.  Done all, PTAL.

Thanks,
Le

File: alf/nest/utils.py [hunk 13]
```diff
@@ -230,8 +230,11 @@ def _traverse_transform(nested, levels):
             return nested._replace(**{level: new_val})
         elif isinstance(nested, dict):
             new_val = nested.copy()
-            new_val[level] = _traverse_transform(
-                nested=nested[level], levels=levels[1:])
+            _val = _traverse_transform(nested=nested[level], levels=levels[1:])
+            if _val is not None:
```
Comment (emailweixu) [inline alf/nest/utils.py]: This is unexpected behavior. What if the caller really want to transform it to None?
And with this behavior, None has different effect for namedtuple and dict. It seems not right.

File: alf/utils/common.py [hunk 14]
```diff
@@ -790,42 +797,107 @@ def detach(nests):
...
         training (bool): True for training, False for unrolling
     """
-    global _training
-    _training = training
+    global _exe_mode
+    _exe_mode = mode
 
 
-def is_training():
+def exe_mode_str():
```
Comment (emailweixu) [inline alf/utils/common.py]: exe_mode_name is a better function name

File: alf/utils/common.py [hunk 15]
```diff
@@ -441,7 +445,10 @@ def forward(self, observation):
         """Normalize a given observation. If during unroll, then first update
         the normalizer. The normalizer won't be updated in other circumstances.
         """
-        if not is_training():
+        # TODO: To update on unroll, we need to fix the case where test exp
+        # should be excluded.
```
Comment (emailweixu) [inline alf/utils/common.py]: test is performed using policy_trainer._step(), which is under eval mode. So there seems no problem.

Comment (le-horizon) [reply alf/utils/common.py]: Done.

File: alf/nest/utils.py (hunk 13 above)
Comment (le-horizon) [reply alf/nest/utils.py]: good point.  reverted.

File: alf/utils/common.py (hunk 14 above)
Comment (le-horizon) [reply alf/utils/common.py]: Done.

**Review by le-horizon — COMMENTED** [review]
Addressed all comments, @hnyu and @emailweixu , could you take another look?

Thanks,
Le

**Review by emailweixu — APPROVED** [review]

**Review by hnyu — APPROVED** [review]
